README.md
# NAME

Amon2::Plugin::Web::Github::Webhook - Github Webhook!

# SYNOPSIS

    use Amon2::Lite;

    __PACKAGE__->load_plugin(
        'Web::Github::Webhook', { allow => '192.30.252.0/22' }
    );

    post '/hook' => sub {
        my $c = shift;
        return $c->res_403 unless my $github = $c->github;

        # $github = {
        #     evnet    => 'push',
        #     delivery => '19967914-9615-11e3-85b5-...',
        #     payload  => {
        #         after => '6fbadfc1ac7b6a7d32bca086ac95d049e0d8b03c',
        #         before => '4f9585dd9fbd86f20c21e3521e4de48a504f58b0',
        #         commits => [ {...} ],
        #         ...,
        #         pusher => ...,
        #         ref => 'refs/heads/master',
        #         repository => {...},
        #     },
        # };

        if ($github->{event} eq 'push') {
            my $payload = $github->{payload};
            my $ref = $payload->{ref};
            ...
        }
        return $c->create_simple_status_page(200 => "OK");
    };

    __PACKAGE__->to_app;

# DESCRIPTION

Amon2::Plugin::Web::Github::Webhook helps you write github webhook
application.

## CONFIGURATION

You can initialize this plugin with access restrictions:

    Your::App->load_plugin(
        "Web::Github::Webhook", { allow => '192.30.252.0/22' }
    );

Default: `192.30.252.0/22`.

You may check the github notification IP addresses for webhooks by:

    curl -i https://api.github.com/meta

## METHOD

This plugin adds `$c->github` method to your application.
If the request is a github notification,
then it returns a hash containing 'event', 'delivery' and 'payload' as in
SYNOPSIS; otherwise returns false.

Here 'the request is a github notification' means that:

    * the request ip address is in the configured ip range.
    * the request has a request header X-GITHUB-EVENT.
    * the request has the body parameter 'payload', and it is a json string.

# SEE ALSO

[Plack::App::GitHub::WebHook](https://metacpan.org/pod/Plack::App::GitHub::WebHook)

[http://developer.github.com/webhooks/](http://developer.github.com/webhooks/)

# LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Shoichi Kaji <skaji@cpan.org>


Changes
Revision history for Perl extension Amon2-Plugin-Web-Github-Webhook

{{$NEXT}}

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Amon2-Plugin-Web-Github-Webhook',
    module_name     => 'Amon2::Plugin::Web::Github::Webhook',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Amon2/Plugin/Web/Github/Webhook.pm
package Amon2::Plugin::Web::Github::Webhook;
use 5.008005;
use strict;
use warnings;

our $VERSION = "0.01";
use JSON ();
use Amon2::Util ();
use Net::CIDR::Lite;

sub init {
    shift;
    my $c      = shift;
    my $config = shift || +{};

    my $allow = $config->{allow} || '192.30.252.0/22';
    my @ip     = ref $allow ? @$allow : ($allow);
    my $github_addr = Net::CIDR::Lite->new;
    $github_addr->add_any($_) for @ip;

    my $decoder = JSON->new->utf8(1);
    Amon2::Util::add_method($c, 'github', sub {
        my $c = shift;

        my $req = $c->req;

        return unless $github_addr->find($req->address);

        return unless my $event   = $req->env->{HTTP_X_GITHUB_EVENT};
        return unless my $payload = $req->body_parameters_raw->{payload};
        return unless my $hash    = eval { $decoder->decode($payload) };

        my $delivery = $req->env->{HTTP_X_GITHUB_DELIVERY} || '';

        return { event => $event, delivery => $delivery, payload => $hash };
    });
}


1;
__END__

=encoding utf-8

=for stopwords webhook webhooks

=head1 NAME

Amon2::Plugin::Web::Github::Webhook - Github Webhook!

=head1 SYNOPSIS

    use Amon2::Lite;

    __PACKAGE__->load_plugin(
        'Web::Github::Webhook', { allow => '192.30.252.0/22' }
    );

    post '/hook' => sub {
        my $c = shift;
        return $c->res_403 unless my $github = $c->github;

        # $github = {
        #     evnet    => 'push',
        #     delivery => '19967914-9615-11e3-85b5-...',
        #     payload  => {
        #         after => '6fbadfc1ac7b6a7d32bca086ac95d049e0d8b03c',
        #         before => '4f9585dd9fbd86f20c21e3521e4de48a504f58b0',
        #         commits => [ {...} ],
        #         ...,
        #         pusher => ...,
        #         ref => 'refs/heads/master',
        #         repository => {...},
        #     },
        # };

        if ($github->{event} eq 'push') {
            my $payload = $github->{payload};
            my $ref = $payload->{ref};
            ...
        }
        return $c->create_simple_status_page(200 => "OK");
    };

    __PACKAGE__->to_app;

=head1 DESCRIPTION

Amon2::Plugin::Web::Github::Webhook helps you write github webhook
application.

=head2 CONFIGURATION

You can initialize this plugin with access restrictions:

    Your::App->load_plugin(
        "Web::Github::Webhook", { allow => '192.30.252.0/22' }
    );

Default: C<192.30.252.0/22>.

You may check the github notification IP addresses for webhooks by:

    curl -i https://api.github.com/meta

=head2 METHOD

This plugin adds C<< $c->github >> method to your application.
If the request is a github notification,
then it returns a hash containing 'event', 'delivery' and 'payload' as in
SYNOPSIS; otherwise returns false.

Here 'the request is a github notification' means that:

    * the request ip address is in the configured ip range.
    * the request has a request header X-GITHUB-EVENT.
    * the request has the body parameter 'payload', and it is a json string.

=head1 SEE ALSO

L<Plack::App::GitHub::WebHook>

L<http://developer.github.com/webhooks/>

=head1 LICENSE

Copyright (C) Shoichi Kaji.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Shoichi Kaji E<lt>skaji@cpan.orgE<gt>

=cut

